File: 2054_two-best-non-overlapping-events.cpp
class Solution {
public:
    int maxTwoEvents(vector<vector<int>>& ev) {
        int sz = ev.size(), mxVal = 0;
        sort(ev.begin(), ev.end(), [](auto& a, auto& b) { return a[0] < b[0]; });
        vector<int> maxSuffix(sz);
        maxSuffix[sz - 1] = ev[sz - 1][2];
        for (int i = sz - 2; i >= 0; --i) maxSuffix[i] = max(ev[i][2], maxSuffix[i + 1]);
        for (int i = 0; i < sz; ++i) {
            int lo = i + 1, hi = sz - 1, idx = -1;
            while (lo <= hi) {
                int mid = lo + (hi - lo) / 2;
                if (ev[mid][0] > ev[i][1]) idx = mid, hi = mid - 1;
                else lo = mid + 1;
            }
            if (idx != -1) mxVal = max(mxVal, ev[i][2] + maxSuffix[idx]);
            mxVal = max(mxVal, ev[i][2]);
        }
        return mxVal;
    }
};

File: 2054_two_best_non_overlapping_events.java
import java.util.*;

class Solution {
    public int maxTwoEvents(int[][] events) {
        int n = events.length, maxSum = 0;
        Arrays.sort(events, (a, b) -> a[0] - b[0]);
        int[] suffixMax = new int[n];
        suffixMax[n - 1] = events[n - 1][2];
        for (int i = n - 2; i >= 0; --i) {
            suffixMax[i] = Math.max(events[i][2], suffixMax[i + 1]);
        }
        for (int i = 0; i < n; ++i) {
            int left = i + 1, right = n - 1, nextIndex = -1;
            while (left <= right) {
                int mid = left + (right - left) / 2;
                if (events[mid][0] > events[i][1]) {
                    nextIndex = mid;
                    right = mid - 1;
                } else {
                    left = mid + 1;
                }
            }
            if (nextIndex != -1) {
                maxSum = Math.max(maxSum, events[i][2] + suffixMax[nextIndex]);
            }
            maxSum = Math.max(maxSum, events[i][2]);
        }
        return maxSum;
    }
}
